### PR TITLE
Remove redundant constraints take 3

### LINKED ISSUE #705: Avoid redundant constraints applying rules that produce one child
When applying a rule produces only one child pattern, we know that the initial constraint subsumes the constraints on the rule; therefore, it is redundant to add the new constraints.

### PR DESCRIPTION
**Pull request by vladciobanu**
Fixes #705 

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by vladciobanu** [comment]
It might be worth it to move the code I added in `Step.Step` to a different module, but I would like a review now that focuses on the changes to the tests.

File: kore/src/Kore/Step/Step.hs [hunk 1]
```diff
@@ -224,8 +229,26 @@ unifyRule
...
+                    unificationSolution requires
+        isPreRedundant <- evaluate predicate
+        let nonRedundantPredicate =
+                case isPreRedundant of
+                    Just True -> unificationSolution
+                    _         -> unificationSolution <> requires
+        Substitution.normalizeExcept nonRedundantPredicate
+
+    -- | ⌈L ∧ φ⌉ ∧ P → Pre
+    makePredicate unificationSolution requires =
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: The comment helps, but the name `makePredicate` is a little vague: what kind of predicate does this make?

File: kore/src/Kore/Step/Step.hs [hunk 2]
```diff
@@ -224,8 +229,26 @@ unifyRule
     let
         RulePattern { requires = ruleRequires } = rule'
         requires' = Predicate.fromPredicate ruleRequires
-    unification' <- Substitution.normalizeExcept (unification <> requires')
+    unification' <- removeRedundancy unification requires'
     return (rule' `Conditional.withCondition` unification')
+  where
+    removeRedundancy unificationSolution requires = do
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: The name `removeRedundancy` is unclear about what is redundant. Maybe, `removeRedundantRequires`?

File: kore/test/Test/Kore/OnePath/Step.hs [hunk 3]
```diff
@@ -312,19 +312,9 @@ test_onePathStrategy =
                     $ Mock.f Mock.b
             ]
         assertEqualWithExplanation ""
-            [ Stuck Conditional
```
Comment (ttuegel) [inline kore/test/Test/Kore/OnePath/Step.hs]: This change does not seem reasonable to me. We cannot unconditionally prove the claim `constr10(b) => a` given only the axiom `constr10(b) => a requires f(b) = c`.

File: kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs [hunk 4]
```diff
@@ -80,18 +82,11 @@ test_definitionEvaluation =
                     { results =
                         OrPattern.fromPattern Conditional
                             { term = Mock.g Mock.a
-                            , predicate = requirement
+                            , predicate = makeTruePredicate
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs]: Where in the pre-condition does it imply that `f(a) = g(b)`?

File: kore/test/Test/Kore/Step/Function/Integration.hs [hunk 5]
```diff
@@ -1028,5 +1032,42 @@ evaluate
...
+env =
+    Env
+        { metadataTools       = tools
+        , simplifierTermLike  = Mock.termLikeSimplifier
+        , simplifierPredicate = Mock.predicateSimplifier
+        , simplifierAxioms    = Mock.axiomSimplifiers
+        }
+  where
+    tools =
+        Mock.makeMetadataTools
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Function/Integration.hs]: IIRC, `Mock.makeMetadataTools` does some things that are sketchy, and I intentionally avoided it for this module. Why is this necessary?

File: kore/test/Test/Kore/Step/Axiom/UserDefined.hs [hunk 6]
```diff
@@ -107,14 +107,14 @@ test_userDefinedFunction =
     , testCase "Cannot apply step with unsat axiom pre-condition" $ do
         let expect =
                 AttemptedAxiom.Applied AttemptedAxiomResults
-                    { results = OrPattern.fromPatterns []
-                    , remainders = OrPattern.fromPatterns
+                    { results = OrPattern.fromPatterns
                         [ Conditional
-                            { term = Mock.functionalConstr10 (mkVar Mock.x)
+                            { term = Mock.functionalConstr11 (mkVar Mock.x)
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Axiom/UserDefined.hs]: The axiom below cannot apply because it `requires False`.

File: kore/test/Test/Kore/Step/Step.hs [hunk 7]
```diff
@@ -546,12 +546,6 @@ test_applyRewriteRule_ =
         actual <- applyRewriteRule_ initial axiomEnsuresBottom
         assertEqualWithExplanation "" expect actual
 
-    , testCase "rule a => b requires \\bottom" $ do
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Step.hs]: Why was this removed?

File: kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs [hunk 8]
```diff
@@ -272,48 +262,6 @@ test_firstFullEvaluation =
                 )
                 (Mock.functionalConstr10 (mkVar Mock.x))
         assertEqualWithExplanation "" expect actual
-    , testCase "Error when not fully rewriting" $ do
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs]: These errors should not have gone away.

File: kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs [hunk 9]
```diff
@@ -352,18 +300,11 @@ test_simplifierWithFallback =
                     { results = OrPattern.fromPatterns
                         [ Conditional
                             { term = Mock.g Mock.a
-                            , predicate = requirement
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Axiom/EvaluationStrategy.hs]: I don't think these tests should have changed, either.

File: kore/src/Kore/Step/Step.hs (hunk 2 above)
Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Step.hs]: Actually, wouldn't `addRequiresUnlessRedundant` be a better name? This does not remove anything...

File: kore/src/Kore/Step/Step.hs [hunk 10]
```diff
@@ -224,8 +229,26 @@ unifyRule
...
+        let predicate =
+                (makePredicate `on` Predicate.toPredicate)
+                    unificationSolution requires
+        isPreRedundant <- evaluate predicate
+        let nonRedundantPredicate =
+                case isPreRedundant of
+                    Just True -> unificationSolution
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Step.hs]: So, I'm a bit confused by this. I think we only implemented the "Just False" case, see https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/SMT/Evaluator.hs#L143
so I would expect this to be a no-op. On the other hand, your change seems to make a difference for the tests below.

So, I'm not sure what comment to make here, but this does not seem right.

**Comment by vladciobanu** [comment]
@ttuegel Yep, you are exactly right. I somehow convinced myself that those were right, but they were obviously not. The problem was that I tried to be cute with `Function.&` and accidentally reversed the implication of the predicate, thus making it `Pre → ⌈L ∧ φ⌉ ∧ P` which turns out to be true pretty often.

Additionally, @virgil-serbanuta was correct: we only have the `Just False` case for `Unsat`, so I needed to negate the predicate anyway.

I rebased/force push-ed the new version which looks a lot better now, I think (no change needed to other tests, added a test that is failing on `master` but working here; can you think of other tests I should add?).

@virgil-serbanuta also had an interesting idea I am inclined to take on: what if we change `evaluate` (this tricked me since it returns `Maybe Bool` but it only really returns `Nothing` or `Just False`) to be more precise? For example, we could add some argument that specifies whether we want to prove something's always true or always false and it would negate the predicate itself if needed...?
